### PR TITLE
mapCursorToPath bug fix

### PR DESCRIPTION
**Pull request by daesunp**
## Description

This PR fixes a bug with an edge case with the use of `moveCursorToPath` in `forestSummarizer.ts`, where when passing the path as `undefined` through this method in chunked forest, the cursor is just cleared (whereas in object forest, it is set to the roots' cursor).

This PR introduces a new method `getCursorAboveDetachedFields`, which returns the cursor at the first node of the chunked forest's roots.

### CODE REVIEW

File: experimental/dds/tree2/src/core/forest/forest.ts [hunk 1]
```diff
@@ -99,7 +100,16 @@ export interface IForestSubscription extends Dependee, ISubscribable<ForestEvent
 	 * This dummy node can be used to read the detached fields,
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/core/forest/forest.ts:100]: This comment needs updating.

File: experimental/dds/tree2/src/core/forest/forest.ts [hunk 2]
```diff
@@ -99,7 +100,16 @@ export interface IForestSubscription extends Dependee, ISubscribable<ForestEvent
...
 	 */
-	moveCursorToPath(destination: UpPath | undefined, cursorToMove: ITreeSubscriptionCursor): void;
+	moveCursorToPath(destination: UpPath, cursorToMove: ITreeSubscriptionCursor): void;
+
+	/**
+	 * The cursor is moved to a special dummy node above the detached fields.
+	 * This dummy node can be used to read the detached fields,
+	 * but other operations (such as inspecting the dummy node's type or path) should not be relied upon.
+	 * As this method does not return a forest cursor and cannot be freed, the caller of this method must
+	 * use with caution and ensure that it is not used incorrectly.
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/core/forest/forest.ts:110]: We can't just tell people not to use it incorrectly. These docs are where we describe how you are allowed to use it. Also these limits are the same as the ones that apply to forest cursors, so its not because this isn't a forest cursors.
```suggestion
	 * While this method does not return an {@link ITreeSubscriptionCursor}, similar restrictions apply to its use:
	 * the returned cursor must not used after any edits are made to the forest.
```

File: experimental/dds/tree2/src/test/snapshots/testTrees.ts [hunk 3]
```diff
@@ -26,7 +27,10 @@ import {
 import { AllowedUpdateType, FieldKey, JsonableTree, UpPath, rootFieldKey } from "../../core";
 import { leaf, SchemaBuilder } from "../../domains";
 
-const factory = new SharedTreeFactory({ jsonValidator: typeboxValidator });
+const factory = new SharedTreeFactory({
+	jsonValidator: typeboxValidator,
+	forest: ForestType.Optimized,
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/snapshots/testTrees.ts:32]: this change seems unrelated to this PR, and I'm not sure why its being made.

File: experimental/dds/tree2/src/core/forest/forest.ts [hunk 4]
```diff
@@ -99,7 +100,16 @@ export interface IForestSubscription extends Dependee, ISubscribable<ForestEvent
...
+	 */
+	getRootCursor(): ITreeCursorSynchronous;
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/core/forest/forest.ts:112]: People might expect this to point to the document root. TO help emphasize that it does not, I think naming it "getCursorAboveDetachedFields" would help.

Comment (CraigMacomber) [inline experimental/dds/tree2/src/core/forest/forest.ts:112]: You are adding a method to the forest API: the forest test suite should be updated to include tests for it.. Apparently it was missing tests for this case before (which enabled this problem in the first place). It would be good to ensure moveCursorToPath has tests for the other cases (that are still supported) as well.

**Comment by CraigMacomber** [comment]
PR description should probably say what the bug was, not just that this fixes one.

Also when fixing bugs, its good practice to include a regression test to show the previously broken case which now works, as well as to ensure that it does not break again.

File: experimental/dds/tree2/src/core/forest/forest.ts (hunk 1 above)
Comment (daesunp) [reply experimental/dds/tree2/src/core/forest/forest.ts]: Updated comment

File: experimental/dds/tree2/src/core/forest/forest.ts (hunk 2 above)
Comment (daesunp) [reply experimental/dds/tree2/src/core/forest/forest.ts]: Updated comment with suggestion

File: experimental/dds/tree2/src/test/snapshots/testTrees.ts (hunk 3 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/snapshots/testTrees.ts]: removed unrelated change

File: experimental/dds/tree2/src/core/forest/forest.ts (hunk 4 above)
Comment (daesunp) [reply experimental/dds/tree2/src/core/forest/forest.ts]: renamed to `getCursorAboveDetachedFields`

Comment (daesunp) [reply experimental/dds/tree2/src/core/forest/forest.ts]: Added unit tests for `moveCursorToPath` and `getCursorAboveDetachedFields`.

I created separate tests for these methods (not as a part of `forestTestSuite.ts`), as the `moveCursorToPath` for the ObjectForest has an additional test case (path set to `undefined`)

File: experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts [hunk 5]
```diff
@@ -86,6 +100,38 @@ describe("ChunkedForest", () => {
...
+
+					const cursor = forest.allocateCursor();
+					const path: UpPath = {
+						parent: undefined,
+						parentField: rootFieldKey,
+						parentIndex: 0,
+					};
+
+					forest.moveCursorToPath(path, cursor);
+					assert.deepEqual(path, cursor.getPath());
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts:118]: we have a helper, expectEqualPaths, that is a bit more robust for doing this (it won't fail if the two paths use different implementations for example). Would be better to use it here, and in the other test below.

File: experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts [hunk 6]
```diff
@@ -86,6 +100,38 @@ describe("ChunkedForest", () => {
 				factory: (schema) => buildChunkedForest(chunker(schema)),
 				skipCursorErrorCheck: true,
 			});
+			describe("moveCursorToPath", () => {
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts:103]: these tests should be added to the forest test suite (which is run just above this) that is applied to all implementations of forest, not independently implemented for each forest. If there is something exter that a specific forest profides, it can have a dedicated test for that, but as long as something is part of the forest interface, the tests should be in the forest test suite to avoid duplication.

File: experimental/dds/tree2/src/test/feature-libraries/objectForest.spec.ts [hunk 7]
```diff
@@ -80,34 +80,18 @@ describe("object-forest", () => {
...
+		forest.moveCursorToPath(undefined, cursor);
+		assert.deepEqual(cursor.getFieldKey(), singleMapTreeCursor(forest.roots).getFieldKey());
 	});
 
 	it("getCursorAboveDetachedFields", () => {
 		const forest = buildForest();
 		initializeForest(forest, [singleJsonCursor([1, 2])]);
 		const cursor = forest.getCursorAboveDetachedFields();
-		assert.deepEqual(cursor.getFieldKey(), singleMapTreeCursor(forest.roots).getFieldKey());
+		assert.deepEqual(cursor, singleMapTreeCursor(forest.roots));
```
Comment (CraigMacomber) [inline experimental/dds/tree2/src/test/feature-libraries/objectForest.spec.ts:95]: deepEqual doesn't really do the correct whing when comparing cursors. This might happen to work right now, but could easily break if for example the implementation changed to a different cursor implementation.

Its better to check if the cursors expose the correct tree content.

For example save the jsonable tree you used to a local, then copy the cursor into a jsonable tree, then compare those two jsonable trees with deep equal rather than comparing the cursor objects.

File: experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts (hunk 6 above)
Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts]: This applies to getCursorAboveDetachedFields as well.

File: experimental/dds/tree2/src/test/feature-libraries/objectForest.spec.ts (hunk 7 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/objectForest.spec.ts]: Updated test to compare jsonableTrees instead of cursors

File: experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts (hunk 6 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts]: moved common `moveCursorToPath` test to the forest test suite

File: experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts (hunk 5 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts]: updated to use expectEqualPaths

File: experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts (hunk 6 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts]: I've kept the getCursorAboveDetachedFields tests away from the forestTestSuite as the chunkedForest implementation steps into the first node of the forest roots, whereas the objectForest implementation just gets the cursor from the forest root (mapTree). Would there be a better way to test for both cases at the same time?

Comment (CraigMacomber) [reply experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts]: These methods are both implementing the same API. You should write tests that they do what the API says they do. The fact that one of them implements that contract using chunked cursors and the other using map cursors should not matter here: they are supposed to both behave according to the API documentation, and if they don't do that, the test should fail and the implementation should be changed.

Comment (daesunp) [reply experimental/dds/tree2/src/test/feature-libraries/chunked-forest/chunkedForest.spec.ts]: Ahh yes, that makes sense. I realized I can just use mapCursorField and convert them into jsonableTrees for comparison. I  have moved the tests to the forestSuite.ts.

**Review by CraigMacomber — APPROVED** [review]
